app/controllers/PayPalOneOff.scala

```diff
@@ -54,21 +57,30 @@ class PayPalOneOff(
       "payerId" -> PayerID
     )
 
-    def processPaymentApiResponse(success: Boolean): Result = {
-      if (success) {
-        SafeLogger.info(s"One-off contribution for Paypal payment is successful. Sending user to thank-you page")
-        Redirect("/contribute/one-off/thankyou")
-      } else {
-        SafeLogger.error(scrub"Error making paypal payment. Sending user to error page.")
-        Ok(views.html.main(
-          "Support the Guardian | PayPal Error",
-          "paypal-error-page",
-          "payPalErrorPage.js",
-          "payPalErrorPageStyles.css"
-        ))
-      }
-    }
-
+    val paypalErrorPage = Ok(views.html.main(
+      "Support the Guardian | PayPal Error",
+      "paypal-error-page",
+      "payPalErrorPage.js",
+      "payPalErrorPageStyles.css"
+    ))
+    def processPaymentApiResponse(paymentApiResponse: Option[PaymentAPIResponse[PayPalError, PaypalExecuteSuccessResponse]]): Result =
+      paymentApiResponse.fold(paypalErrorPage)(resp => {
+        resp.data.fold(
+          resp.error.fold(paypalErrorPage)(error => {
+            if (error.errorName.contains("PAYMENT_ALREADY_DONE")) {
+              SafeLogger.info(s"PAYMENT_ALREADY_DONE error code received. Sending user to thank-you page")
+              Redirect("/contribute/one-off/thankyou")
```

Comment:
Can this be [reverse-routed](https://www.playframework.com/documentation/2.6.x/ScalaRouting#Reverse-routing)?